Cabal/Distribution/Simple/GHC.hs

```diff
@@ -962,11 +964,56 @@ gbuildModDefFiles (GReplExe  _)     = []
 gbuildModDefFiles (GBuildFLib flib) = foreignLibModDefFile flib
 gbuildModDefFiles (GReplFLib  flib) = foreignLibModDefFile flib
 
+-- | "Main" module name when overridden by @ghc-options: -main-is ...@
+-- or 'Nothing' if no @-main-is@ flag could be found.
+--
+-- In case of 'Nothing', 'Distribution.ModuleName.main' can be assumed.
+exeMainModuleName :: Executable -> Maybe ModuleName
+exeMainModuleName Executable{buildInfo = bnfo} =
+    -- GHC honors the last occurence of a module name updated via -main-is
+    --
+    -- Moreover, -main-is when parsed left-to-right can update either
+    -- the "Main" module name, or the "main" function name, or both,
+    -- see also 'decodeMainIsArg'.
+    msum $ reverse $ map decodeMainIsArg $ findIsMainArgs ghcopts
+  where
+    ghcopts = hcOptions GHC bnfo
+
+    findIsMainArgs [] = []
+    findIsMainArgs ("-main-is":arg:rest) = arg : findIsMainArgs rest
+    findIsMainArgs (_:rest) = findIsMainArgs rest
+
+-- | Decode argument to '-main-is'
+--
+-- Returns 'Nothing' if argument set only the function name.
+--
+-- This code has been stolen/refactored from GHC's DynFlags.setMainIs function
+decodeMainIsArg :: String -> Maybe ModuleName
+decodeMainIsArg arg
+  | not (null main_fn) && isLower (head main_fn)
+                        -- The arg looked like "Foo.Bar.baz"
+  = Just (ModuleName.fromString main_mod)
```

Comment:
Do we actually check that `main_mod` is a valid module name? The way I see it, it could be something like `Foo.b$r.baz`, and this code would still accept it. I'd use `simpleParse` instead of `ModuleName.fromString` here.

Another possible refactoring is to add a type `QualifiedName = QualifiedName { qualModName :: ModuleName, qualName :: String }` that could potentially be also useful elsewhere, then this function would become just ``msum [ qualModName `fmap` simpleParse arg, simpleParse arg ]`` (assuming that `QualifiedName` has a `Text` instance).